drivers/usb/udc/udc_ambiq.c

```diff
@@ -0,0 +1,915 @@
+#include <soc.h>
+#include "am_mcu_apollo.h"
+#include <string.h>
+#include <zephyr/drivers/clock_control/clock_control_ambiq.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/usb/udc.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/cache.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/irq.h>
+#include "udc_common.h"
+
+LOG_MODULE_REGISTER(udc_ambiq, CONFIG_UDC_DRIVER_LOG_LEVEL);
+
+/* UDC event ID enumeration */
+enum udc_ambiq_event_type {
+	/* SETUP packet received at Control Endpoint */
+	UDC_AMBIQ_EVT_HAL_SETUP,
+	/* OUT transaction completed */
+	UDC_AMBIQ_EVT_HAL_OUT_CMP,
+	/* IN transaction completed */
+	UDC_AMBIQ_EVT_HAL_IN_CMP,
+	/* Xfer request received via udc_ambiq_ep_enqueue API */
+	UDC_AMBIQ_EVT_XFER,
+};
+
+/* UDC event queue element structure */
+struct udc_ambiq_event {
+	const struct device *dev;
+	enum udc_ambiq_event_type type;
+	uint8_t ep;
+};
+
+/* Define message queue for UDC events */
+K_MSGQ_DEFINE(drv_msgq, sizeof(struct udc_ambiq_event), CONFIG_UDC_AMBIQ_MAX_QMESSAGES,
+	      sizeof(void *));
+
+/* USB device controller access from devicetree */
+#define DT_DRV_COMPAT ambiq_usb
+
+/* Max Packet Size for USB endpoints */
+#define EP0_MPS   64U
+#define EP_FS_MPS 64U
+#define EP_HS_MPS 512U
```

Comment:
Later the code sets the iso capability on the endpoints. Are the endpoints limited in hardware to these specific values and not the maximum allowed values?

Reply:
Hi @tmon-nordic, there is a limitation that the end of transaction is not able to be identified when OUT endpoint received data that is multiple of endpoint MPS, when the data count is lower than the triggered size. Hence, the safest way to expose this now is to limit it to 64 bytes for FS and 512 bytes for HS.

Reply:
Where is this limitation, in your shim driver, HAL driver, or controller?

Reply:
I think it can be categorized as controller limitation.